UCSC-Keysight/OpenTAP-Cobot-Plugin
# Feature/listener

# Feature/listener, Implement OpenTAP ResultListener

## Summary
Branch implements an OpenTAP `resultListener` that catalogs data collected from the Universal Robots' controller unit in a `.csv` file. 

## Details

### Files of Interest
1. `MoveCobot.py`
     - added `PublishResult()` to support the listener. 
3. `CsvResultListener.py`, 
    - implements `resultListener` object.

### Testing Instructions
1. Launch the UR3e simulator with the following command:
    ```
    docker run -it -e ROBOT_MODEL=UR3e -p 30001:30001 -p 30002:30002 -p 30004:30004 -p 6080:6080 --name ur3e_container universalrobots/ursim_e-series
    ```
2. Perform the following commands to launch the Keysight editor:
      
    ```Console
    pip install tabulate
    git clone https://github.com/UCSC-Keysight/OpenTAP-Cobot-Plugin.git
    cd OpenTAP-Cobot-Plugin/openTap
    git checkout -b feature/listener origin/feature/listener
    dotnet build
    bin\tap editor
    ```
5. Once in the editor, go to File > Open then select `openTap/PullRequest.TapPlan`. Ensure the `UR3e` instrument is using your IPv4 then run the test plan. 
 
### Expected Behavior / Output

At the conclusion of the test plan, a passing verdict will be given and a file, `OpenTAP-Cobot-Plugin\openTap\bin\MyFile.csv`, will be generated with similar contents as the following:

![image](https://user-images.githubusercontent.com/80125540/233522087-7812fc79-b821-4a80-8006-9542ef90ce8d.png)

### Miscellaneous
- Minor modifications were made to `.gitignore` to better constrain the file tree.
- Minor modifications were made to `MoveCobot.py` unrelated to the `ResultListener` for readability / efficiency purposes.
- Minor modifications were made to `README.md` to make documentation more accurate.
